### PR TITLE
Embedding visualization is added to TensorBoard callback.

### PR DESCRIPTION
**Pull request by smyskoff**
Hi.

I was using Keras for my task, fitting a network with embedding layers. I wanted to discover the progress of my network, just to feel that things goes as I expect, and I found such a thing as TensorBoard, which helps us to visualize training process. I wanted to visualize my embeddings conveniently, and I found that TensorBoard can do this, but I did not find this ability in a TensorFlow callback for Keras. Not a big deal, I made some changes in a code and now it's possible to visualize embeddings from the box for me.

So, let me introduce you my changes and whether these changes can be helpful for others.

I'm not a pro in TensorFlow, so I could did something wrong. I used [this](https://www.tensorflow.org/how_tos/embedding_viz/) documentation page, and wrote quite a bit of code, so I hope even though my code is not totally correct, it's easy to fix.

## My use case
I do not know the way others want use such a feature, but let me describe how I make it.

### How often
Parameter `embeddings_freq` governs how often an embeddings checkpoint is saved.

### What to visualize
The way I tell the callback which layers to visualize is by a list of layer names `embeddings_layer_names`. I govern layer names when I create a model, so it was quite a natural way for me to do this.

### Understanding visualization
Next thing is the labels for what we visualized.

In my case there's just integral numeric data is on the embedding layer input, so there's no way to deal with just visualized points in 3D with indices as labels. I think it's general case, as that's the way embeddings work as far as I know them. I have dictionary which maps embedding layer input indices to a real labels behind them, so I store metadata files and pass their names to a constructor of the callback. `embeddings_metadata` is a dictionary which maps layer names to a file with corresponding metadata.

_NOTE:_ I have not try images visualization yet.

### Progress visualization
Frankly, I was disappointed in some way when I'd done, as I expected to see embeddings progress in dynamics. I thought this way as in TensorBoard documentation I used I saw:
> Periodically save your embeddings in a `LOG_DIR`.

and I did not get it. When the training is done I can find the file named `checkpoint` in `LOG_DIR`. It looks like this:
```
model_checkpoint_path: "embedding_data.ckpt-594"
all_model_checkpoint_paths: "embedding_data.ckpt-198"
all_model_checkpoint_paths: "embedding_data.ckpt-297"
all_model_checkpoint_paths: "embedding_data.ckpt-396"
all_model_checkpoint_paths: "embedding_data.ckpt-495"
all_model_checkpoint_paths: "embedding_data.ckpt-594"
```
I can change `model_checkpoint_path` to any of present `all_model_checkpoint_paths`, but that's it, there's only one of this checkpoints will be visualized at the same time. It's not a question or issue report to you, just my feelings.

screenshot from 2017-02-02 00-17-08

### Concerns
Even though I wanted to save just one from bunch of embedding layers in the model, I can visualize other layers on an Embeddings tab on TensorBoard Page.

Yep, by default I see precisely what I wanted to see, but I can select any layer from my network there from dropdown box.

Inspecting the `LOG_DIR` I see all saved «checkpoints», and it looks like they are just a saved model, not only the tensor I wanted to save. Also theirs file size tells me that they are just dumped model.

screenshot from 2017-02-02 00-17-13

## Another thing to consider
Just another thing I did not make changes for, but want to discuss is a usage of frequencies in this callback. Epochs in Keras log are shown in one-based form, i.e. they are counted from one. In opposite, `on_epoch_end()` method gets zero-based epoch index. What's inconvenience for me is that if I do training in 40 epochs with frequency of saving histograms and embeddings of 10, I expect to visualize 10-th, 20-th, 30-th and 40-th, but it comes to visualization of 1-th (I'm not sure it does make sense), 11-th, 21-th and 31-th. Finally I «missed» 9 epochs.

Isn't it more natural to increment `epoch` parameter in `on_batch_end` before checking whether anything have to be logged?

### CODE REVIEW

**Comment by xiaoleihuang** [comment]
Thanks for the update. I used the pip install the keras, but I did not find the feature available on my keras ('1.2.2') ?
BTW, how to use your code to visualize the embeddings? I am not clear with the parameter: embeddings_layer_names.
When we add the pretrained embedding layers, there seems no name, just use model.add(pretrain_embedding). And when we define the Embedding, no name option there?

Thank you.

**Comment by smyskoff** [comment]
Hello.

Thank you for your response.

> I used the pip install the keras, but I did not find the feature available on my keras ('1.2.2') ?

Regarding installation using `pip` try this:
```sh
[darth.vader@deathstar ~]$ virtualenv-3 venv
[darth.vader@deathstar ~]$ source venv/bin/activate
[darth.vader@deathstar ~]$ pip install git+https://github.com/smyskoff/keras.git
```

I checked—this brings you my version of repo.

> BTW, how to use your code to visualize the embeddings?

I don't see concise way to show a snippet which will work in terms of copy-paste, as it's a model with an input preprocessed data, but the idea is next:

```python
# Input data parameters and desired embeddings size.
dictionary_size = 100
embeddings_size = 20
num_inputs = 3

inputs = []
embeddings = []

# Prepare several inputs for CBOW.
for i in range(num_inputs):
    inp = Input(shape=(1,), dtype='int32', name='input_{}'.format(i))
    layer = Embedding(input_dim=dictionary_size, output_dim=embeddings_size,
                      input_length=1, name='embeddings_{}'.format(i))
    # 3D tensor to 2D for further merge.
    layer = Flatten()(layer)
    inputs.append(inp)
    embeddings.append(layer)

# Ready to merge all the embedding layers together.
layer = merge(embeddings, mode='concat')

# Put some layers on top. Names here are option as we don't care
# of this layers, so they'll be default ones in case if we omit this parameter.
layer = Dense(embeddings_size * num_inputs, activation='relu', name='hidden_1')(layer)
layer = Dense(embeddings_size * num_inputs, activation='relu', name='hidden_2')(layer)
layer = Dense(embeddings_size * num_inputs, activation='relu', name='hidden_3')(layer)

# Output layer has to have the same dimensions as input embeddings for CBOW.
# Also get a probabilistic output with softmax activation function.
layer = Dense(dictionary_size, activation='softmax', name='output')(layer)

model = Model(input=inputs, output=layer)

# Whatever you want here.
model.compile(...)

# Prepare data.
X = ...
Y = ...

# Metadata file consist of newline-separated words, which are associated
# with corresponding input data index. For example it can look like this:
# --- metadata file begin ---
# super
# cool
# metadata
# file
# --- metadata file end ---
#
# For this example dictionary is expected to be be of size 4, for which
# 0 is associated with the word 'super', 1 -> 'cool', 2 -> 'metadata' and 3 -> 'file'.
# You'll see this metadata on the projector's page.

# Keep eye on all our embedding layers.
embeddings_to_monitor = ['embeddings_{}'.format(i)
                         for i in range(num_inputs)]
metadata_file_name = 'metadata.tsv'
embeddings_metadata = {layer_name: metadata_file_name
                       for layer_name in embeddings_to_monitor}

tb_callback = TensorBoard(histogram_freq=10, write_graph=False,
                          embeddings_freq=100,                          # Store each 100 epochs...
                          embeddings_layer_names=embeddings_to_monitor, # this list of embedding layers...
                          embeddings_metadata=embeddings_metadata)      # with this metadata associated with them.

# Ready to fit.
model.fit(X, Y, np_epoch=1000, batch_size=512, callbacks=[tb_callback])
```

Also take a look at screenshots I attached to the issue description to see how embedding visualization looks on TensorBoard page. For the moment 'Embeddings' tab on the TensorBoard page is empty, with my patch you can see 3D projection of the high-dimensional embeddings.

> BTW, how to use your code to visualize the embeddings? I am not clear with the parameter: embeddings_layer_names.
> When we add the pretrained embedding layers, there seems no name, just use model.add(pretrain_embedding). And when we define the Embedding, no name option there?

I see few ways to deal with layer names (frankly, I don't know how to identify them in the other way):

- when you create a layer to add to a model, you pass the parameter `name='layer_name'` to a constructor;
- when you've got pre-trained model, it's loaded with the layers' names when you use `model.save()` API;
- if you've just got a Numpy array weights, you can initialize a newly created layer with this array using `weights=[your_pre_trained_weights]` constructor parameter. Also note that Embedding layer does not have biases, so there's only one Numpy array in the initialization list.

My patch is also survey to know how community use Keras and how to make the feature useful, so this is opened question.

**Comment by gewoonrik** [comment]
I think you have to PR this to keras-2 instead of master.

**Comment by smyskoff** [comment]
There's no problem to merge my changes into keras-2 branch.

Is there any point to open new PR for the other branch? Maybe we can confirm with changes in this (forum) branch, but include changes to whatever (git) branch it can be useful and appropriate.

**Comment by gewoonrik** [comment]
hmm, Keras 2 was released yesterday, so I think master is fine now.

**Comment by fchollet** [comment]
I will try to review this shortly. Anyone else willing to, can help review as well.

File: keras/callbacks.py [hunk 1]
```diff
@@ -605,6 +616,31 @@ def __init__(self, log_dir='./logs',
         self.merged = None
         self.write_graph = write_graph
         self.write_images = write_images
+        self.embeddings_freq = embeddings_freq
+        self.embeddings_layer_names = embeddings_layer_names
```
Comment (tboquet) [inline keras/callbacks.py]: by default you could maybe add all embedding layers' names by checking the type of each layer in the model so that it doesn't affect the naming when building the model. You could then bypass the creation of a `tf.train.Saver` by checking if `embeddings_freq` is not 0.

File: keras/callbacks.py [hunk 2]
```diff
@@ -635,6 +671,14 @@ def set_model(self, model):
         else:
             self.writer = tf.summary.FileWriter(self.log_dir)
 
+        self.saver = tf.train.Saver()
```
Comment (tboquet) [inline keras/callbacks.py]: Should be conditional to `embeddings_freq` from here.

File: keras/callbacks.py [hunk 3]
```diff
@@ -635,6 +671,14 @@ def set_model(self, model):
...
+        self.embeddings = {layer.name: (layer.weights[0],
+                                        os.path.join(self.log_dir, layer.name +
+                                                     '.ckpt'))
+                           for layer in self.model.layers
+                           if layer.name in self.embeddings_layer_names}
+        self.embeddings_metadata_setup()
```
Comment (tboquet) [inline keras/callbacks.py]: not sure you need to wrap this in a method it looks clean this way though.

**Comment by tboquet** [comment]
This is a really useful feature! I tried the callback which is working correctly aside from the observations made by @smyskoff.
A more general observation: this callback seems to perform a lot of different things now, should it be worth looking into a way to split parts that are independent to the validation data.

**Comment by smyskoff** [comment]
Corrected test script:

```python
import numpy as np

from keras.models import Model
from keras.layers import Input, Dense, Flatten, Embedding
from keras.layers.merge import concatenate
from keras.callbacks import TensorBoard
from keras.utils.np_utils import to_categorical

dictionary_size = 100
embeddings_size = 20
num_inputs = 3

inputs = []
embeddings = []

# Prepare several inputs for CBOW.
for i in range(num_inputs):
    inp = Input(shape=(1,), dtype='int32', name='input_{}'.format(i))
    layer = Embedding(input_dim=dictionary_size, output_dim=embeddings_size,
                      input_length=1, name='embeddings_{}'.format(i))(inp)
    # 3D tensor to 2D for further merge.
    layer = Flatten()(layer)
    inputs.append(inp)
    embeddings.append(layer)

# Ready to merge all the embedding layers together.
layer = concatenate(embeddings)

# Put some layers on top. Names here are option as we don't care
# of this layers, so they'll be default ones in case if we omit this parameter.
layer = Dense(embeddings_size * num_inputs, activation='relu',
              name='hidden_1')(layer)
layer = Dense(embeddings_size * num_inputs, activation='relu',
              name='hidden_2')(layer)
layer = Dense(embeddings_size * num_inputs, activation='relu',
              name='hidden_3')(layer)

# Output layer has to have the same dimensions as input embeddings for CBOW.
# Also get a probabilistic output with softmax activation function.
layer = Dense(dictionary_size, activation='softmax', name='output')(layer)

model = Model(inputs=inputs, outputs=layer)

# Whatever you want here.
model.compile(optimizer='rmsprop', loss='categorical_crossentropy',
              metrics=['accuracy'])

num_samples = 1000

X = [np.random.randint(dictionary_size, size=num_samples) for _ in
     range(num_inputs)]
Y = to_categorical(np.random.randint(dictionary_size, size=num_samples),
                   dictionary_size)

embeddings_layer_names = None # or some list of layers ['embeddings_1']

tb_callback = TensorBoard(histogram_freq=10, write_graph=False,
                          embeddings_freq=100,
                          embeddings_layer_names=embeddings_layer_names)

# Ready to fit.
model.fit(X, Y, epochs=400, batch_size=512, callbacks=[tb_callback])

```

File: keras/callbacks.py [hunk 4]
```diff
@@ -671,19 +673,25 @@ def set_model(self, model):
...
-        self.embeddings = {layer.name: (layer.weights[0],
-                                        os.path.join(self.log_dir, layer.name +
-                                                     '.ckpt'))
-                           for layer in self.model.layers
-                           if layer.name in self.embeddings_layer_names}
-        self.embeddings_metadata_setup()
+        if self.embeddings_freq:
+            self.saver = tf.train.Saver()
+
+            embeddings_layer_names = self.embeddings_layer_names \
```
Comment (tboquet) [inline keras/callbacks.py]: Could you use multiple lines instead of a broken one liner?

File: keras/callbacks.py [hunk 5]
```diff
@@ -671,19 +673,25 @@ def set_model(self, model):
...
+                                     if self.embeddings_layer_names else \
+                                        [layer.name for layer in self.model.layers
+                                         if type(layer).__name__ == 'Embedding']
+            self.embeddings = {layer.name:
```
Comment (tboquet) [inline keras/callbacks.py]: Same here, not sure if the dict comprehension is readable.

File: keras/callbacks.py [hunk 6]
```diff
@@ -671,19 +673,25 @@ def set_model(self, model):
...
+                               for layer in self.model.layers
+                               if layer.name in embeddings_layer_names}
+            self.embeddings_metadata_setup()
 
     def on_epoch_end(self, epoch, logs=None):
         logs = logs or {}
 
         if self.validation_data and self.histogram_freq:
-            if epoch % self.histogram_freq == 0:
+            if (epoch + 1) % self.histogram_freq == 0:
```
Comment (tboquet) [inline keras/callbacks.py]: why do you add 1 here?

Comment (smyskoff) [inline keras/callbacks.py]: As I described in the initial post, `epoch` initial value is 0, not 1, so `epoch` modulo `self.embeddings_freq` for initial epoch is 0. It can be unexpected behaviour. For example, I asked to dump embeddings each 10 epochs, and train model for 20 ones. There will be dumped two epochs, 0-th and 10-th, but I expect 9-th and 19-th to be dumped (i.e. 10-th and 20-th in one-based reckoning).

File: keras/callbacks.py (hunk 5 above)
Comment (smyskoff) [inline keras/callbacks.py]: I tried to fit 80 symbols width keeping code readable. Yeap, mapping is splitted into two lines and it looks not pretty nice. Sorry, I did not read code conventions for the project, and I don't know whether it appropriate to make long lines.

Regarding loop and conditional parts—I placed them on different lines to be easily distinguished.

File: keras/callbacks.py (hunk 4 above)
Comment (smyskoff) [inline keras/callbacks.py]: I can re-write an expression not to make it in one line. Something like this:

```python
embeddings_layer_names = self.embeddings_layer_names

if not embeddings_layer_names:
    embeddings_layer_names = [layer.name for layer in self.model.layers
                              if type(layer).__name__ == 'Embedding']
```

But in such an expression I used there's no parenthesis, so new line character is interpreted as the end of expression.

Also I tried to fit in 80 characters width.

**Comment by smyskoff** [comment]
Also here's CI log, frankly I have no idea how should I put indentations. I'm using Emacs with `python-mode`, and it makes indentations automatically, so I don't change anything in general.

```log
__________________ PEP8-check(ignoring E501 \ E402 \ E731 \) ___________________
/home/travis/build/fchollet/keras/keras/callbacks.py:633:16: E127 continuation line over-indented for visual indent
               {layer_name: self.embeddings_metadata
               ^
/home/travis/build/fchollet/keras/keras/callbacks.py:680:38: E127 continuation line over-indented for visual indent
                                     if self.embeddings_layer_names else \
                                     ^
/home/travis/build/fchollet/keras/keras/callbacks.py:681:41: E127 continuation line over-indented for visual indent
                                        [layer.name for layer in self.model.layers
                                        ^
/home/travis/build/fchollet/keras/keras/callbacks.py:684:17: E128 continuation line under-indented for visual indent
                (layer.weights[0], os.path.join(self.log_dir,
                ^
/home/travis/build/fchollet/keras/keras/callbacks.py:686:32: E126 continuation line over-indented for hanging indent
                               for layer in self.model.layers
                               ^
============================= 330 tests deselected =============================
```

File: keras/callbacks.py (hunk 4 above)
Comment (tboquet) [inline keras/callbacks.py]: Yes please, rewrite those lines.

File: keras/callbacks.py (hunk 5 above)
Comment (tboquet) [inline keras/callbacks.py]: It's possible to write this one liner in several more readable lines of code. It will also solve your pep8 problems.

File: keras/callbacks.py (hunk 6 above)
Comment (tboquet) [inline keras/callbacks.py]: Please remove these changes so that this callback has the same behavior as the other callbacks. It could be discussed in another issue/PR.

**Comment by tboquet** [comment]
It's possible `python-mode` does not point out some pep8 errors. Take a look at https://github.com/proofit404/anaconda-mode.

File: keras/callbacks.py (hunk 1 above)
Comment (smyskoff) [inline keras/callbacks.py]: Done

**Comment by gokceneraslan** [comment]
@smyskoff which version of tf is required for the projector, 0.12? It would be nice to inform users (e.g. see https://github.com/fchollet/keras/issues/6545) somehow about the required version of tf, it might be tricky since it's not written in setup.py and on the web page.

**Comment by smyskoff** [comment]
Tried to upgrade TensorFlow using `pip3`, actual TensorFlow version is 1.1.0:

```bash
$ pip3 freeze | grep -e tensorflow -e Keras
Keras==2.0.4
tensorflow-gpu==1.1.0
```

Cannot tell which is the minimum TensorFlow version to make it work.

Also checked with `import keras` and with the posted script—for me works just fine.

BTW looks like problem in referred thread is solved.
